iceoryx_examples/ara_com/ara/com/method_client.hpp

```diff
@@ -0,0 +1,115 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_ARA_COM_METHOD_CLIENT_HPP
+#define IOX_EXAMPLES_ARA_COM_METHOD_CLIENT_HPP
+
+#include "iceoryx_hoofs/cxx/optional.hpp"
+#include "iceoryx_posh/popo/client.hpp"
+
+#include "ara/types.hpp"
+
+#include <memory>
+#include <thread>
+#include <utility>
+
+namespace ara
+{
+namespace com
+{
+class MethodClient
+{
+  public:
+    MethodClient(const core::String& service, const core::String& instance, const core::String& event)
+        : m_client({service, instance, event})
+    {
+        m_waitset.attachState(m_client, iox::popo::ClientState::HAS_RESPONSE).or_else([](auto) {
+            std::cerr << "failed to attach client" << std::endl;
+            std::exit(EXIT_FAILURE);
+        });
+    }
+
+    ~MethodClient()
+    {
+        m_waitset.detachState(m_client, iox::popo::ClientState::HAS_RESPONSE);
+    }
+
+    ara::com::Future<AddResponse> operator()(uint64_t addend1, uint64_t addend2)
+    {
+        bool requestSuccessfullySend{false};
+        m_client.loan()
+            .and_then([&](auto& request) {
+                request.getRequestHeader().setSequenceId(m_sequenceId);
+                request->addend1 = addend1;
+                request->addend2 = addend2;
+                request.send().and_then([&]() { requestSuccessfullySend = true; }).or_else([](auto& error) {
+                    std::cout << "Could not send Request! Error: " << error << std::endl;
+                });
+            })
+            .or_else([&](auto& error) {
+                std::cout << "Could not allocate Request! Error: " << error << std::endl;
+                requestSuccessfullySend = false;
+            });
+
+        if (!requestSuccessfullySend)
+        {
+            return ara::com::Future<AddResponse>();
+        }
+
+        Promise<AddResponse> promise;
+        auto future = promise.get_future();
+        std::thread(
+            [&](Promise<AddResponse>&& promise) {
+                auto notificationVector = m_waitset.timedWait(iox::units::Duration::fromSeconds(5));
+
+                for (auto& notification : notificationVector)
+                {
+                    if (notification->doesOriginateFrom(&m_client))
+                    {
+                        while (m_client.take().and_then([&](const auto& response) {
+                            auto receivedSequenceId = response.getResponseHeader().getSequenceId();
+                            if (receivedSequenceId == m_sequenceId)
+                            {
+                                AddResponse result{response->sum};
+                                m_sequenceId++;
+                                promise.set_value_at_thread_exit(result);
```

Comment:
I only could move it one line earlier after having created `result`.

Reply:
I mean not set it at thread exit with `set_value`. The point is to make the result available as fast as possible. In this case this makes little difference since the thread will exit almost immediately.